### ISSUE TITLE
bulletproof-font-face : false positive when no .eot

### ISSUE THREAD

**Issue by axel3rd** [SonarCommunity/sonar-css#120]
Hi, 

This snippet code raises an issue [bulletproof-font-face](http://nemo.sonarqube.org/coding_rules#rule_key=css%3Abulletproof-font-face) :
```
@font-face {
    font-family: 'PT Serif Caption';
    font-style: normal;
    font-weight: 400;
    src: local('PT Serif Caption'), local('PTSerif-Caption'), url(../font/myFont.woff)
        format('woff');
}
```
I understand the issue, the fix to apply ... but if I don't care of IE and don't want provide ``myFont.eot`` (it is evil, yes, but ok), why this issue ?

I'm not a CSS specialist and perhaps the problem comes from my misunderstanding about ``@font-face``.

Best regards.

**Comment by racodond** [comment]
If you don't care about IE and more specifically IE 6, 7 and 8, you should deactivate this "Font face should be made compatible with IE 6, 7 and 8" rule. This being said, I'm not sure to understand your point. You'd like this rule to be deactivated by default (that I would agree with by the way)?

**Comment by axel3rd** [comment]
@racodond : Thanks for the reply. 

The desactivation is out of scope, beyond the *"don't care"* (it was for the sample), some of my projects on the SonarQube platform must support historical IE.

This rule has sense when an ``.eot`` is found in ``src``, must be first with the fix (and could stay by default, always true if ``.eot`` presents). 

The fact to have always ``.eot``/``.woff``/``.ttf``/``.svg`` formats in ``src`` (=> all browsers supported) is IMO another rule scope (this rule activated by default would be in this case a little tricky).

**Comment by racodond** [comment]
> The desactivation is out of scope, beyond the "don't care" (it was for the sample), some of my projects on the SonarQube platform must support historical IE.

OK

> This rule has sense when an .eot is found in src, must be first with the fix (and could stay by default, always true if .eot presents).

OK. But let's get back to your example:
`src: local('PT Serif Caption'), local('PTSerif-Caption'), url(../font/myFont.woff) format('woff');`
On IE <9, the browser will first try to find a local font 'PT Serif Caption' installed on your computer. Let's say it doesn't find it. It will then try to find a local font 'PTSerif-Caption'. Let's say, it doesn't find it. It will then try to find a font at URL `../font/myFont.woff) format('woff'`. That will return a 404. Thus it is not working fine on IE < 9. Hence the issue that is raised. Am I missing something?


> The fact to have always .eot/.woff/.ttf/.svg formats in src (=> all browsers supported) is IMO another rule scope (this rule activated by default would be in this case a little tricky).

And not really valuable from my point of view

**Comment by axel3rd** [comment]
Sorry for the delay, it was an "out of connection" weekend for me.

> On IE <9, the browser will first tryv [...] Hence the issue that is raised. Am I missing something?

No I'm agree with you, but this rule could have currently two different read ... because the rule description is focalized on the 404 bug behavior and the way to fix it (very valuable) ; not on the fact you need `eot` format for IE<9 support.

1) The one you explain to me : a `@font-face` should support IE<9 in ever case.

=> The rule description could explicit that `eot` format is required for IE<9 support, and the snippet like [this link](http://blog.fontspring.com/2011/04/best-practices-for-serving-webfonts-to-ie9/) (all IE supported with IIS, ...) ; and in this case, I'm agree with your first post, this rule could be deactivated by default (IE<9 is less 5% of browsers).

2) The one I have when I see the rule description & snippet code :

The issue is raised on :
```
@font-face {
    font-family: 'MyWebFont';
    src: url('webfont.woff') format('woff');
}
```

When I see the rule description :

> When using @font-face to declare multiple font types for cross browser compatibility, 

Heu ... this is not my case, I use only a single source *woff*, supported by modern browser (the rule doesn't say : you *must* support IE by using an `eot` format).

> Check that the first file is the .eot file and that the workaround for IE is set
> Noncompliant : src: url('myfont-webfont.eot')
> Compliant :  src: url('myfont-webfont.eot?#iefix')

It would have been very useful if I had used `eot` in a bad way (not first, without fix) ... but I don't use this format in my snippet.

=> The value added of this rule is only when `eot` is one of format used, and could stay by default.

(The difference is subtle, I know, but I was not the alone to try to understand this rule in my company ... I'm just the devil advocate here :-).)

**Comment by racodond** [comment]
@axel3rd I didn't forget about it. I need to find some time to dig into it to really understand the issue and the workaround. I'll keep you posted.

**Comment by axel3rd** [comment]
@racodond No problem. Thanks.

**Comment by racodond** [comment]
@axel3rd @kalidasya I dove into this `@font-face` topic and now better understand it. I propose to deprecate this rule and create a new one: `@font-face` should be made compatible with the required browsers.

It is based on https://css-tricks.com/snippets/css/using-font-face/
The new rule has one parameter: browser support. It can be set to:

* Practical Level of Browser Support
* Slightly Deeper Browser Support
* Deepest Possible Browser Support


##### Deepest Browser Support
```
@font-face {
  font-family: 'MyWebFont';
  src: url('webfont.eot'); /* IE9 Compat Modes */
  src: url('webfont.eot?#iefix') format('embedded-opentype'), /* IE6-IE8 */
       url('webfont.woff2') format('woff2'), /* Super Modern Browsers */
       url('webfont.woff') format('woff'), /* Pretty Modern Browsers */
       url('webfont.ttf')  format('truetype'), /* Safari, Android, iOS */
       url('webfont.svg#svgFontName') format('svg'); /* Legacy iOS */
}
```
The rule checks that each `@font-face` containing `url` function in `src` properties:

* Defines two `src` properties
* First `src` property defining only one `.eot` file with no `format` function
* Second `src` property with `.eot` file defined first with the `?#iefix` hack
* Second `src` property with all formats defined in the above order
* `local` are not allowed in the `src` properties

##### Slightly Deeper Browser Support
```
@font-face {
  font-family: 'MyWebFont';
  src: url('myfont.woff2') format('woff2'),
       url('myfont.woff') format('woff'),
       url('myfont.ttf') format('truetype');
}
```
The rule checks that each `@font-face` containing `url` function in `src` property defines at least `woff2` and `woff`  and `truetype` formats in this order. Additional format can be defined but only after those `woff2` and `woff` formats. `local` can be defined as well in the `src` property. An exception though, an `.eot` file can be defined before the `woff2` format. Another `src` property defining an `.eot` file only can be defined before this `src` property. Basically, if a `Deepest Browser Support` syntax is found, it should not raise any issue.

##### Practical Level of Browser Support
```
@font-face {
  font-family: 'MyWebFont';
  src:  url('myfont.woff2') format('woff2'),
        url('myfont.woff') format('woff');
}
```
The rule checks that each `@font-face` containing `url` function in `src` property defines at least `woff2` and `woff` formats in this order. Additional format can be defined but only after those `woff2` and `woff` formats. `local` can be defined as well in the `src` property. Another `src` property defining an `.eot` file only can be defined before this `src` property. Basically, if a `Deepest Browser Support` syntax is found, it should not raise any issue.

**Comment by axel3rd** [comment]
*Sorry for the delay, it was holidays time.*

@racodond : I'm globally agree with your previous purpose.

Minor remark : According chapter [Understanding Font File Types - WOFF / WOFF2](https://css-tricks.com/snippets/css/using-font-face/) ; WOFF2 format improves compression file **but not** the browser type support => out of scope for this rule IMO. If WOFF is here, the wanted browser type support is achieved.

**Comment by racodond** [comment]
> Minor remark : According chapter Understanding Font File Types - WOFF / WOFF2 ; WOFF2 format improves compression file but not the browser type support => out of scope for this rule IMO. If WOFF is here, the wanted browser type support is achieved.

It would be fully linked to browser support if "Super Progressive Browser Support" was implemented. I didn't want to implement this browser support level for now as it is not very useful for now. Nevertheless, even if it's not fully in the scope, I'd like to keep WOFF2 in this rule as it is a good habit to start supporting new browser capabilities.

Here's a PR: https://github.com/SonarCommunity/sonar-css/pull/132
Feedback is welcome!

**Comment by axel3rd** [comment]
> Nevertheless, even if it's not fully in the scope, I'd like to keep WOFF2 in this rule as it is a good habit to start supporting new browser capabilities.

ok :-)
